File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.py
class Solution:
    def letterCombinations(self, digits: str) -> List[str]:
        letter_map = {
            "2": "abc",
            "3": "def",
            "4": "ghi",
            "5": "jkl",
            "6": "mno",
            "7": "pqrs",
            "8": "tuv",
            "9": "wxyz",
        }
        ans,curr=[],[]

        def formCombi(ans,curr,ind,letter_map,digits):
            if ind==len(digits):
                ans.append("".join(curr))
                return
            
            for i in letter_map[digits[ind]]:
                
                curr.append(i)

                formCombi(ans,curr,ind+1,letter_map,digits)
                curr.pop()
        formCombi(ans,curr,0,letter_map,digits)
        
        return ans

    
        

File: 0017-letter-combinations-of-a-phone-number/0017-letter-combinations-of-a-phone-number.cpp
class Solution {
public:
    void generateCombi(map<char,vector<char>>&mp,int ind, string curr,string &digits,vector<string>&res){
        if(curr.length() == digits.length()){
            res.push_back(curr);
            return;
        }
        for(char ch : mp[digits[ind]]){
            curr+=ch;
            generateCombi(mp,ind+1,curr,digits,res);
            curr.pop_back();
        }
        
    }
    
    vector<string> letterCombinations(string digits) {

        map<char,vector<char>>mp;
        string curr = "";
        vector<string>res;
        mp['2']={'a','b','c'};
        mp['3']={'d','e','f'};
        mp['4']={'g','h','i'};
        mp['5']={'j','k','l'};
        mp['6']={'m','n','o'};
        mp['7']={'p','q','r','s'};
        mp['8']={'t','u','v'};
        mp['9']={'w','x','y','z'};

        generateCombi(mp,0,curr,digits,res);
        return res;
              
    }
};